src/kudu/consensus/leader_election.cc

```diff
@@ -138,6 +138,113 @@ bool VoteCounter::AreAllVotesIn() const {
+  FetchTopologyInfo(voter_distribution_, data_quorum_requirement_);
+  for (auto regional_vote_count : voter_distribution_) {
+    CHECK_GT(regional_vote_count.second, 0);
+    num_voters_ += regional_vote_count.second;
+    yes_vote_count_.emplace(regional_vote_count.first, 0);
+    no_vote_count_.emplace(regional_vote_count.first, 0);
+  }
+
+  CHECK_GT(num_voters_, 0);
+
+  bool non_empty_quorum = false;
+  for (auto regional_quorum_count : data_quorum_requirement_) {
+    auto total_voters = voter_distribution_.find(regional_quorum_count.first);
+    CHECK(total_voters != voter_distribution_.end());
+    CHECK_LE(regional_quorum_count.second, total_voters->second);
+    if (regional_quorum_count.second > 0) {
+      non_empty_quorum = true;
+    }
+  }
+  CHECK(non_empty_quorum);
+
+  for (const RaftPeerPB& peer : config_.peers()) {
+    if (peer.member_type() == RaftPeerPB::VOTER) {
+      uuid_to_region_.emplace(peer.permanent_uuid(), peer.attrs().region());
+    }
+  }
+}
+
+Status FlexibleVoteCounter::RegisterVote(
+    const std::string& voter_uuid, ElectionVote vote, bool* is_duplicate) {
+  Status s = VoteCounter::RegisterVote(voter_uuid, vote, is_duplicate);
+  if (ContainsKey(uuid_to_region_, voter_uuid)) {
+    const std::string region = uuid_to_region_[voter_uuid];
+    switch (vote) {
+      case VOTE_GRANTED:
+        InsertIfNotPresent(&yes_vote_count_, region, 0);
+        yes_vote_count_[region]++;
+        break;
+      case VOTE_DENIED:
+        InsertIfNotPresent(&no_vote_count_, region, 0);
+        no_vote_count_[region]++;
+        break;
+    }
+  } else {
+    return Status::InvalidArgument(
+        Substitute("UUID {$0} not present in config.", voter_uuid));
+  }
+  return s;
+}
+
+std::pair<bool, bool> FlexibleVoteCounter::GetQuorumState() const {
+  bool quorum_satisfied = true;
+  bool quorum_satisfaction_possible = true;
+  for (auto regional_vote_count : data_quorum_requirement_) {
+    const string region = regional_vote_count.first;
+
+    auto regional_yes_count = yes_vote_count_.find(region);
+    auto regional_no_count = no_vote_count_.find(region);
+    auto total_region_count = voter_distribution_.find(region);
+
+    CHECK(regional_yes_count != yes_vote_count_.end());
+    CHECK(regional_no_count != no_vote_count_.end());
+    CHECK(total_region_count != voter_distribution_.end());
+
+    if (regional_yes_count->second < regional_vote_count.second) {
+      quorum_satisfied = false;
+    }
+    if (regional_no_count->second + regional_vote_count.second >
+        total_region_count->second) {
+      quorum_satisfaction_possible = false;
+    }
+  }
+  return std::make_pair(quorum_satisfied, quorum_satisfaction_possible);
+}
+
+bool FlexibleVoteCounter::IsDecided() const {
+  auto quorum_state = GetQuorumState();
+  return quorum_state.first || !quorum_state.second;
+}
+
+Status FlexibleVoteCounter::GetDecision(ElectionVote* decision) const {
+  auto quorum_state = GetQuorumState();
```

Comment:
seems like we will call the same expensive function for IsDecided and GetDecision

Reply:
const reference